eclipse_plugin/src/com/google/googlejavaformat/java/AbstractJavaFormatter.java

```diff
@@ -0,0 +1,162 @@
+          createSnippetFormatter()
+              .format(
+                  snippetKind, source, rangesFromRegions(regions), initialIndent, includeComments);
+      if (idempotent(source, regions, replacements)) {
+        // Do not create edits if there's no diff.
+        return null;
+      }
+      // Convert replacements to text edits.
+      return editFromReplacements(replacements);
+    } catch (IllegalArgumentException | FormatterException exception) {
+      // Do not format on errors.
+      return null;
+    }
+  }
+
+  private List<Range<Integer>> rangesFromRegions(IRegion[] regions) {
+    List<Range<Integer>> ranges = new ArrayList<>();
+    for (IRegion region : regions) {
+      ranges.add(Range.closedOpen(region.getOffset(), region.getOffset() + region.getLength()));
+    }
+    return ranges;
+  }
+
+  /** @return {@code true} if input and output texts are equal, else {@code false}. */
+  private boolean idempotent(String source, IRegion[] regions, List<Replacement> replacements) {
+    // This implementation only checks for single replacement.
+    if (replacements.size() == 1) {
+      Replacement replacement = replacements.get(0);
+      String output = replacement.getReplacementString();
+      // Entire source case: input = output, nothing changed.
+      if (output.equals(source)) {
+        return true;
+      }
+      // Single region and single replacement case: if they are equal,
+      // nothing changed.
+      if (regions.length == 1) {
+        Range<Integer> range = replacement.getReplaceRange();
+        String snippet = source.substring(range.lowerEndpoint(), range.upperEndpoint());
+        if (output.equals(snippet)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  private TextEdit editFromReplacements(List<Replacement> replacements) {
+    // Split the replacements that cross line boundaries.
+    TextEdit edit = new MultiTextEdit();
+    for (Replacement replacement : replacements) {
+      Range<Integer> replaceRange = replacement.getReplaceRange();
+      edit.addChild(
+          new ReplaceEdit(
+              replaceRange.lowerEndpoint(),
+              replaceRange.upperEndpoint() - replaceRange.lowerEndpoint(),
+              replacement.getReplacementString()));
+    }
+    return edit;
+  }
+
+  /**
+   * Create an instance of the formatter that will be used. This method should be aligned with
+   * {@link #indentationSize()} such that the indentation size corresponds with the indent level of
+   * the format options used to create the snippet formatter.
+   */
+  protected abstract SnippetFormatter createSnippetFormatter();
+
+  /**
+   * The number of spaces per indent level. This method should be aligned with {@link
+   * #createSnippetFormatter()} such that the indentation size corresponds with the indent level of
+   * the format options used to create the snippet formatter.
+   */
+  protected abstract int indentationSize();
```

Comment:
🤔 maybe `SnippetFormatter#indentationSize()` can be made `public` instead? That does mean a modification to its public API, but since this caller could use it, maybe so can others.

Reply:
I can make it public, though I tend to prefer more restrictive access until a need arises. That said, I'll make change if you prefer. Please advise.

Reply:
Well, it seems a need arose ;).

But in all seriousness: I'm not a maintainer of this project, so my advice carries little weight. Let's see what the maintainers say.